### PR TITLE
test: improve annotation pipeline test coverage (#62)

### LINKED ISSUE #62: test: improve annotation pipeline test coverage to catch contig/AF parsing issues
## Background

Issues #58 and #60 were both production bugs discovered during real `genechat init` runs on Fly.io that the existing test suite did not catch. This issue analyzes why the gaps existed and proposes concrete test improvements.

### Issue #58: Multi-allelic gnomAD AF crash
`float()` failed on comma-separated AF values like `"0.25,."` from multi-allelic sites. The `update_gnomad_from_stream` method passed raw AF strings to `float()` without handling `Number=A` VCF fields.

### Issue #60: gnomAD annotation writes 0 variants (contig name mismatch)
SnpEff may strip the `chr` prefix from contig names (storing `chrom='1'`), while bcftools gnomAD output preserves the `chr` prefix (`chrom='chr1'`). The UPDATE WHERE clause didn't match, silently writing 0 variants.

Both bugs were fixed (PRs #59, #61), but the test gaps that allowed them to ship remain structurally present. The fixes added targeted regression tests for the specific failure modes, but the underlying testing patterns that permitted these classes of bugs have not been addressed.

---

### Problem 1: The test VCF contains no multi-allelic sites

`scripts/generate_test_vcf.py` defines 26 synthetic variants — all biallelic with simple scalar AF values like `AF=0.14`. Real genomes (including GIAB NA12878) contain thousands of multi-allelic sites where gnomAD's `Number=A` fields produce comma-separated values like `0.25,.`.

The `ensure_test_patch_db` fixture in `tests/conftest.py` builds the test patch.db by reading the test VCF through pysam, which correctly decomposes tuples. But even if it had used raw text streams, the test data contained no multi-allelic AF values to trigger the `float()` crash.

### Problem 2: The test fixture uses a single-source stream for all annotation layers

`ensure_test_patch_db` (lines 27-85 of `tests/conftest.py`) feeds the *same* `ann_lines` list to all three stream methods:
```python
db.populate_from_snpeff_stream(iter(ann_lines))
db.update_clinvar_from_stream(iter(ann_lines))
db.update_gnomad_from_stream(iter(ann_lines))
```

Because these lines originate from the same VCF (read via pysam), all three streams share identical contig names (`chr1`, `chr12`, etc.). In production, each annotation step uses a *different* tool's output:
- **SnpEff** may output bare names (`1`, `2`, ...)
- **bcftools** (ClinVar, gnomAD, dbSNP) preserves the input VCF's `chr` prefix

The fixture never tests the cross-tool contig mismatch scenario.

### Problem 3: No "pipeline simulation" test exercises the multi-step annotation sequence

The unit tests in `test_patch.py` test each `update_*_from_stream` method in isolation. Most tests populate with SnpEff first, then update with gnomAD/ClinVar, which is correct — but all use consistently-prefixed contig names. There is no test that simulates the realistic pipeline sequence where different tools produce different contig formats.

### Problem 4: The e2e tests don't verify annotation data completeness

The GIAB e2e tests in `tests/e2e/test_vcf_engine_giab.py` verify genotype zygosity and variant structure but never assert that `population_freq["global"]` is non-None. Test line 124 checks `"population_freq" in v` (key existence) but not that it contains actual values. A genome with 0 gnomAD annotations would pass all e2e tests.

### Problem 5: No edge-case AF values in test data

The `test_update_gnomad` test (line 160 of `test_patch.py`) uses `AF=0.14;AF_grpmax=0.21` — clean scalar values. The tests added by PR #59 (`test_update_gnomad_multiallelic_af`, `test_update_gnomad_all_missing_af`, `test_parse_af_helper`) are good regression tests but were written *after* the bug. The original test data contained no adversarial AF patterns.

---

### 1. Add multi-allelic variants to the test VCF generator

**File:** `scripts/generate_test_vcf.py`

Add 2-3 variants with multi-allelic AF values to the `VARIANTS` list. Since `Number=A` means one value per ALT allele, a biallelic record from gnomAD can still carry multi-allelic AF strings when bcftools transfers annotations from a multi-allelic gnomAD site. Example additions:

```python
# Multi-allelic AF from gnomAD (biallelic user site, multi-allelic gnomAD source)
("chr14", 94844947, "rs12879035", "G", "A", ".", "PASS",
 "ANN=A|missense_variant|MODERATE|SERPINA1|...;AF=0.03,.;AF_popmax=.,0.05",
 "0/1"),

# All-missing AF (gnomAD has no data for any allele)
("chr17", 7674220, "rs28934578", "G", "C", ".", "PASS",
 "ANN=C|missense_variant|MODERATE|TP53|...;AF=.,.;AF_popmax=.",
 "0/1"),
```

This ensures `ensure_test_patch_db` exercises `_parse_af()` on every test run.

### 2. Simulate cross-tool contig mismatch in `ensure_test_patch_db`

**File:** `tests/conftest.py`

Modify `ensure_test_patch_db` to simulate the realistic annotation pipeline where SnpEff strips the `chr` prefix but bcftools preserves it:

```python
# For SnpEff stream: strip chr prefix to simulate SnpEff behavior
snpeff_lines = []
for line in ann_lines:
    if line.startswith("chr"):
        line = line[3:]  # "chr1\t..." -> "1\t..."
    snpeff_lines.append(line)

db.populate_from_snpeff_stream(iter(snpeff_lines))
# ClinVar and gnomAD streams keep chr prefix (bcftools behavior)
db.update_clinvar_from_stream(iter(ann_lines))
db.update_gnomad_from_stream(iter(ann_lines))
```

This would have caught issue #60 immediately — the gnomAD UPDATE would have matched 0 rows.

### 3. Add a dedicated "pipeline simulation" test class

**File:** `tests/test_patch.py`

Add a test class that exercises the full write-then-read cycle with realistic contig mismatches and multi-allelic values, without requiring external tools:

```python
class TestAnnotationPipelineSimulation:
    """Simulate the full annotation pipeline with realistic tool output differences.
    
    Each annotation step in production uses a different external tool that may
    produce different contig naming conventions. These tests verify the complete
    pipeline sequence works end-to-end.
    """

    def test_full_pipeline_cross_format_contigs(self, tmp_path):
        """SnpEff bare contigs + bcftools chr-prefixed contigs -> all data present."""
        db = PatchDB.create(tmp_path / "pipeline.db")
        
        # Step 1: SnpEff outputs bare contig names
        snpeff_lines = [
            "1\t100\t.\tA\tG\t.\tPASS\tANN=G|missense_variant|MODERATE|GENE1||\tGT\t0/1\n",
            "12\t21178615\t.\tT\tC\t.\tPASS\t"
            "ANN=C|missense_variant|MODERATE|SLCO1B1|ENSG|transcript|ENST|protein_coding||c.521T>C|p.Val174Ala||||||\t"
            "GT\t0/1\n",
        ]
        snpeff_count = db.populate_from_snpeff_stream(iter(snpeff_lines))
        assert snpeff_count == 2
        
        # Step 2: bcftools ClinVar outputs chr-prefixed names
        clinvar_lines = [
            "chr12\t21178615\t.\tT\tC\t.\tPASS\t"
            "CLNSIG=drug_response;CLNDN=Simvastatin;CLNREVSTAT=reviewed\tGT\t0/1\n"
        ]
        clinvar_count = db.update_clinvar_from_stream(iter(clinvar_lines))
        assert clinvar_count == 1  # Must match despite different prefix
        
        # Step 3: bcftools gnomAD outputs chr-prefixed names with multi-allelic AF
        gnomad_lines = [
            "chr1\t100\t.\tA\tG\t.\tPASS\tAF=0.03,.;AF_grpmax=.,0.05\tGT\t0/1\n",
            "chr12\t21178615\t.\tT\tC\t.\tPASS\tAF=0.14;AF_grpmax=0.21\tGT\t0/1\n",
        ]
        gnomad_count = db.update_gnomad_from_stream(iter(gnomad_lines))
        assert gnomad_count == 2  # Must match despite different prefix
        
        # Step 4: dbSNP backfill with chr-prefixed names
        dbsnp_lines = [
            "chr1\t100\trs999\tA\tG\t.\tPASS\t.\tGT\t0/1\n",
            "chr12\t21178615\trs4149056\tT\tC\t.\tPASS\t.\tGT\t0/1\n",
        ]
        dbsnp_count = db.update_dbsnp_from_stream(iter(dbsnp_lines))
        assert dbsnp_count == 2
        
        # Verify all data is present and readable
        ann1 = db.get_annotation("chr1", 100, "A", "G")
        assert ann1 is not None
        assert ann1["af"] == pytest.approx(0.03)  # parsed from "0.03,."
        assert ann1["af_grpmax"] == pytest.approx(0.05)  # parsed from ".,0.05"
        assert ann1["rsid"] == "rs999"
        
        ann2 = db.get_annotation("chr12", 21178615, "T", "C")
        assert ann2 is not None
        assert ann2["gene"] == "SLCO1B1"
        assert ann2["clnsig"] == "drug_response"
        assert ann2["af"] == pytest.approx(0.14)
        assert ann2["rsid"] == "rs4149056"
        
        db.close()

    def test_pipeline_reports_zero_updates_on_mismatch(self, tmp_path):
        """Detect if a contig format mismatch causes silent zero-update.
        
        This is a canary test: if normalize_chrom or _chrom_variants_fixed
        regresses, this test fails with gnomad_count == 0.
        """
        db = PatchDB.create(tmp_path / "mismatch.db")
        
        # Populate with bare contigs (simulating SnpEff)
        db._conn.execute(
            "INSERT INTO annotations (chrom, pos, ref, alt, gene) "
            "VALUES ('1', 500, 'A', 'G', 'TEST')"
        )
        db._conn.commit()
        
        # Update with chr-prefixed contigs (simulating bcftools gnomAD)
        gnomad_lines = [
            "chr1\t500\t.\tA\tG\t.\tPASS\tAF=0.10;AF_grpmax=0.15\tGT\t0/1\n"
        ]
        count = db.update_gnomad_from_stream(iter(gnomad_lines))
        assert count == 1, (
            f"gnomAD update matched {count} rows (expected 1). "
            "Contig normalization may have regressed — see issue #60."
        )
        db.close()
```

### 4. Add gnomAD data completeness assertion to e2e tests

**File:** `tests/e2e/test_vcf_engine_giab.py`

Add a test that verifies gnomAD annotations are actually populated, not just structurally present:

```python
class TestAnnotationCompleteness:
    """Verify annotation layers were actually populated, not just structurally present."""

    def test_gnomad_data_populated(self, giab_engine):
        """At least some variants in a well-covered region should have AF data.
        
        Catches issue #60 where gnomAD annotation silently wrote 0 variants.
        """
        # BRCA1 region — well covered by gnomAD exomes
        variants = giab_engine.query_region("chr17:43044295-43170245")
        if not variants:
            pytest.skip("No variants in BRCA1 region")
        
        has_af = [
            v for v in variants
            if v.get("population_freq", {}).get("global") is not None
        ]
        assert len(has_af) > 0, (
            f"Found {len(variants)} variants in BRCA1 but none have gnomAD AF. "
            "gnomAD annotation may have silently failed (see issue #60)."
        )
```

### 5. Add `_extract_info_field` tests for multi-allelic AF patterns

**File:** `tests/test_patch.py`

The existing `TestExtractInfoField` class should include test cases for the multi-allelic AF patterns that triggered issue #58:

```python
def test_multiallelic_af_value(self):
    """Multi-allelic AF values are extracted as raw comma-separated strings."""
    assert _extract_info_field("AF=0.25,.", "AF") == "0.25,."

def test_multiallelic_af_grpmax(self):
    assert _extract_info_field("AF=0.1;AF_grpmax=.,0.30", "AF_grpmax") == ".,0.30"

def test_all_missing_af(self):
    assert _extract_info_field("AF=.,.;AC=5", "AF") == ".,."
```

These verify that `_extract_info_field` faithfully passes through multi-allelic values (the parsing responsibility belongs to `_parse_af`, but extraction must not mangle the raw strings).

### 6. Add `parse_vcf_stream` test with bare contig names

**File:** `tests/test_patch.py`

The existing `TestParseVcfStream` class tests all have `chr`-prefixed input. Add a test with bare contig names to verify normalization works at the parser level:

```python
def test_bare_chrom_normalized(self):
    """Bare contig names (no chr prefix) are normalized correctly."""
    lines = ["1\t100\trs123\tA\tG\t.\tPASS\tAF=0.1\tGT\t0/1\n"]
    results = list(parse_vcf_stream(self._make_stream(lines), ["AF"]))
    assert results[0]["chrom"] == "1"  # Already bare, should stay bare

def test_mixed_chrom_formats(self):
    """Mixed chr-prefixed and bare contig names both normalize."""
    lines = [
        "chr1\t100\t.\tA\tG\t.\tPASS\t.\tGT\t0/1\n",
        "2\t200\t.\tC\tT\t.\tPASS\t.\tGT\t0/1\n",
    ]
    results = list(parse_vcf_stream(self._make_stream(lines), []))
    assert results[0]["chrom"] == "1"
    assert results[1]["chrom"] == "2"
```

---

## Summary of Gaps and Fixes

| Gap | Would have caught | Proposed fix | Effort |
|-----|-------------------|-------------|--------|
| No multi-allelic AF in test VCF | #58 | Add multi-allelic variants to `generate_test_vcf.py` | Small |
| Test fixture uses same contig format for all layers | #60 | Modify `ensure_test_patch_db` to strip chr prefix for SnpEff stream | Small |
| No pipeline simulation test | #58, #60 | Add `TestAnnotationPipelineSimulation` class | Medium |
| e2e tests don't verify gnomAD data completeness | #60 | Add `test_gnomad_data_populated` to e2e suite | Small |
| No multi-allelic AF in `_extract_info_field` tests | #58 | Add 3 test cases to `TestExtractInfoField` | Trivial |
| No bare-contig test in `parse_vcf_stream` | #60 | Add 2 test cases to `TestParseVcfStream` | Trivial |

All proposed tests run without external tools (SnpEff, bcftools, gnomAD downloads) and fit into the standard `uv run pytest` workflow.

## Priority

Items 1, 2, and 3 are the highest impact — they would have caught both bugs before they shipped. Items 4-6 are lower effort and provide defense-in-depth.

### PR DESCRIPTION
**Pull request by natecostello**
## Summary
- Add 2 multi-allelic AF variants to the test VCF generator (`scripts/generate_test_vcf.py`) to exercise `_parse_af()` on every test run
- Modify `ensure_test_patch_db` fixture to simulate cross-tool contig mismatch: SnpEff stream uses bare contigs (`1`, `12`), ClinVar/gnomAD streams use chr-prefixed (`chr1`, `chr12`)
- Add `TestAnnotationPipelineSimulation` class (3 tests) covering the full multi-step pipeline with realistic contig format differences and multi-allelic AF values
- Add gnomAD data completeness assertion to e2e tests, multi-allelic AF extraction tests to `TestExtractInfoField`, and bare-contig normalization tests to `TestParseVcfStream`
- Always regenerate test VCF and rebuild stale patch.db via fingerprint check (Copilot review fixes)
- Use `Number=.` for AF headers in test VCF to accommodate multi-allelic gnomAD transfers

## Test plan
- [x] `uv run pytest --ignore=tests/e2e` — all 373 tests pass
- [x] `uv run ruff check .` — clean
- [x] New `TestAnnotationPipelineSimulation` tests verify cross-tool contig mismatch is handled
- [x] New `TestExtractInfoField` tests verify multi-allelic AF string extraction
- [x] New `TestParseVcfStream` tests verify bare-contig normalization
- [x] e2e `test_gnomad_data_populated` asserts AF data presence in BRCA1 region (skips when gnomAD unavailable)

Closes #62

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: tests/e2e/test_vcf_engine_giab.py [hunk 1]
```diff
@@ -104,6 +104,26 @@ def test_empty_region_returns_empty(self, giab_engine):
...
+    def test_gnomad_data_populated(self, giab_engine):
+        """At least some variants in a well-covered region should have AF data.
+
+        Catches issue #60 where gnomAD annotation silently wrote 0 variants.
+        """
+        # BRCA1 region — well covered by gnomAD exomes
+        variants = giab_engine.query_region("chr17:43044295-43170245")
+        if not variants:
+            pytest.skip("No variants in BRCA1 region")
+
```
[the remark this answers is not in the document]
Comment (natecostello) [reply tests/e2e/test_vcf_engine_giab.py]: Applied this suggestion in commit 4777a9e: added `annotation_versions()` check so the test skips when no gnomAD annotations are configured in the e2e setup.

🤖 Generated with [Claude Code](https://claude.ai/code)

File: scripts/generate_test_vcf.py [hunk 2]
```diff
@@ -393,6 +393,33 @@
...
+        "chr14",
+        94844947,
+        "rs12879035",
+        "G",
+        "A",
+        ".",
+        "PASS",
+        "ANN=A|missense_variant|MODERATE|SERPINA1|ENSG00000197249|transcript|ENST00000440909|protein_coding||c.863A>T|p.Glu288Val||||||;"
+        "AF=0.03,.;AF_popmax=.,0.05",
+        "0/1",
```
[the remark this answers is not in the document]
Comment (natecostello) [reply scripts/generate_test_vcf.py]: Applied this suggestion in commit f6524d6: changed AF, AF_popmax, and AF_grpmax headers from `Number=A` to `Number=.` so the multi-allelic comma-separated values are spec-valid on single-ALT records.

🤖 Generated with [Claude Code](https://claude.ai/code)

File: tests/conftest.py [hunk 3]
```diff
@@ -30,7 +30,15 @@ def ensure_test_patch_db(ensure_test_vcf):
...
-        return
+        # Rebuild if VCF changed (e.g. new variants added) to avoid stale data
+        from genechat.patch import PatchDB
+
+        existing = PatchDB(patch_path, readonly=True)
+        fingerprint_ok = existing.check_vcf_fingerprint(vcf_path)
+        existing.close()
+        if fingerprint_ok:
+            return
+        patch_path.unlink()  # stale — rebuild below
```
[the remark this answers is not in the document]
Comment (natecostello) [reply tests/conftest.py]: Fixed in commit 381ea8d. Now explicitly checks `get_vcf_fingerprint() is not None` before trusting `check_vcf_fingerprint()`. Legacy patch.db files without fingerprints are treated as stale and rebuilt.

🤖 Generated with [Claude Code](https://claude.ai/code)
